dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskDispatchFailedQueueConsumer.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.dolphinscheduler.server.master.config.MasterConfig;
+import org.apache.dolphinscheduler.server.master.metrics.TaskMetrics;
+import org.apache.dolphinscheduler.service.queue.TaskFailedRetryPriority;
+import org.apache.dolphinscheduler.service.queue.TaskPriority;
+import org.apache.dolphinscheduler.service.queue.TaskPriorityQueue;
+
+import java.util.Objects;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.PostConstruct;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Component
+public class TaskDispatchFailedQueueConsumer extends BaseDaemonThread {
+
+    /**
+     * taskPriorityQueue
+     */
+    @Autowired
+    private TaskPriorityQueue<TaskPriority> taskPriorityQueueImpl;
+
+    /**
+     * taskDispatchFailedQueue
+     */
+    @Autowired
+    private TaskPriorityQueue<TaskFailedRetryPriority> taskDispatchFailedQueueImpl;
+
+    @Autowired
+    private MasterConfig masterConfig;
+
+    private ThreadPoolExecutor retryConsumerThreadPoolExecutor;
+
+    protected TaskDispatchFailedQueueConsumer() {
+        super("TaskDispatchFailedQueueConsumerThread");
+    }
+
+    @PostConstruct
+    public void init() {
+        this.retryConsumerThreadPoolExecutor = (ThreadPoolExecutor) ThreadUtils
+                .newDaemonFixedThreadExecutor("TaskDispatchFailedQueueConsumerThread", masterConfig.getDispatchTaskNumber());
+        super.start();
+    }
+
+    @Override
+    public void run() {
+        while (!ServerLifeCycleManager.isStopped()) {
+            try {
+                dispatchFailedBackToTaskPriorityQueue(masterConfig.getDispatchTaskNumber());
+            } catch (Exception e) {
+                TaskMetrics.incTaskDispatchError();
+                log.error("failed task retry error", e);
+            }
+        }
+    }
+
+    /**
+     * put the failed dispatch task into the dispatch queue again
+     */
+    private void dispatchFailedBackToTaskPriorityQueue(int fetchTaskNum) {
+        for (int i = 0; i < fetchTaskNum; i++) {
+            try {
+                TaskFailedRetryPriority delayTaskPriority = taskDispatchFailedQueueImpl.poll(Constants.SLEEP_TIME_MILLIS, TimeUnit.MILLISECONDS);
+                if (Objects.isNull(delayTaskPriority)) {
+                    continue;
+                }
+                retryConsumerThreadPoolExecutor.submit(() -> {
+                    TaskPriority retryTaskPriority = delayTaskPriority.getTaskPriority();
+                    retryTaskPriority.setDispatchFailedRetryTimes(retryTaskPriority.getDispatchFailedRetryTimes() + 1);
+                    taskPriorityQueueImpl.put(retryTaskPriority);
+                });
+            } catch (InterruptedException exception) {
+                log.error("dispatch failed queue poll error", exception);
+                Thread.currentThread().interrupt();
```

Comment:
If the thread interrupted, you need to break the upper while loop.

Reply:
ok, thx.